src/rust/engine/graph/src/entry.rs

```diff
@@ -62,10 +62,25 @@ pub(crate) enum EntryState<N: Node> {
   // The previous_result value is _not_ a valid value for this Entry: rather, it is preserved in
   // order to compute the generation value for this Node by comparing it to the new result the next
   // time the Node runs.
+  //
+  // A note on dirty as was_dirty:
+  // dirty and was_dirty have different meanings for each variant.
+  // Completed::dirty means "The current value may not longer be relevant - you should check whether
+  //   its dependencies have new values, and if so, you should re-compute this value."
+  // Running::dirty means "When this completes, its computed value should be marked dirty
+  //   (and then see Completed::dirty's comment)
+  // NotStarted::was_dirty and Running::was_dirty mean "The _previous_ value (not the current value)
+  //   may no longer be relevant - you should check whether its dependencies have new values, and if
+  //   so, you should re-compute this value, BUT ALSO, the graph containing this node may show edges
+  //   which are no longer valid (and may cause a spurious cycle to be detected if these spurious
+  //   edges happen to form a cycle). Accordingly, if a cycle is detected involving a node which is
+  //   marked was_dirty, you should consider clearing it, removing any edges from it, and re-running
+  //   it.
```

Comment:
This feels a lot like we should break out more enum variants instead of adding a bool field, which might reduce the need for documentation here. It also means these conditionals must be handled in any match statement instead of an `if` that could be forgotten or invalidated after further edits.

Separately, the existing note on the `Running` case:
> // The `previous_result` value for a Running node is not a valid value. See NotStarted.

also seems like it would benefit from splitting these out into more cases, with any necessary common fields in sub-`struct`s.

Reply:
@stuhood discussed this in a comment sent one minute before this one: https://github.com/pantsbuild/pants/pull/7583#discussion_r276375119.

Reply:
Reworked